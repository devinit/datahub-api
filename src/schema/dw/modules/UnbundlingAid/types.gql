type AidUnit {
  # this will usually be a summed up aggregate value
  value: Float
  # country or organisation or channel or bundle name
  name: String
}
type IdNamePair {
  id: String
  name: String
}

type UnbundlingAidSelections {
  # unbundling aid selection options
  toCountries: [IdNamePair]
  fromCountriesAndOrgs: [IdNamePair]
  channels: [IdNamePair]
  sectors: [IdNamePair]
  form: [IdNamePair]
}

input UnbundlingAidQuery {
    # oda or oof 
    aidType: String!
    years: [Int]!
    # eg channel, bundle
    groupBy: String!
    toCountry: String
    fromCountryOrOrg: String
    sector: String
    form: String
    channel: String
}

type Query {
  getUnbundlingAidData(args: UnbundlingAidQuery):[AidUnit]
  getUnbundlingSelectionData: UnbundlingAidSelections
}

