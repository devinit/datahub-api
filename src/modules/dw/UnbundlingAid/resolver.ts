export default {
    Query: {
        async unbundlingAidData(_root, {args}, ctx) {
            return ctx.modules.unbundlingAid.getUnbundlingAidData(args);
        },
        async unbundlingAidDataTotal(_root, {args}, ctx) {
            return ctx.modules.unbundlingAid.getUnbundlingAidDataTotal(args);
        },
        async unbundlingSelectionData(_root, args, ctx) {
            return ctx.modules.unbundlingAid.getUnbundlingSelectionData(args);
        }
    }
};
