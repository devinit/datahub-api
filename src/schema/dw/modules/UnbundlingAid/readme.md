Unbundling aid ODA and OOF schema