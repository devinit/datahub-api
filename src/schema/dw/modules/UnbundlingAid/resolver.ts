export default {
    Query: {
        async unbundlingAidData(_root, {args}, ctx) {
            return ctx.dw.unbundlingAid.getUnbundlingAidData(args);
        },
        async UnbundlingAidToTalQuery(_root, {args}, ctx) {
            return ctx.dw.unbundlingAid.getUnbundlingAidDataTotal(args);
        },
        async unbundlingSelectionData(_root, args, ctx) {
            return ctx.dw.unbundlingAid.getUnbundlingSelectionData(args);
        }
    }
};
