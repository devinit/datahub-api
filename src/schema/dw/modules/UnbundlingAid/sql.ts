export default {
    // tslint:disable-next-line:max-line-length
    total: 'SELECT SUM(value) FROM ( SELECT SUM(value) as value FROM ${table^} where year = ${year} GROUP BY to_di_id) t'
};
