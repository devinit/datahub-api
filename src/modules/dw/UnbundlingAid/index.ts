
import {IDB} from '@devinit/graphql-next/lib/db';
import {makeSqlAggregateQuery, entitesFnMap, DONOR, RECIPIENT, MULTILATERAL, CROSSOVER} from '../../utils';
import {getConceptAsync, IConcept, getConcepts} from '../../refs/concept';
import * as shortid from 'shortid';
import sql from './sql';
import {IEntity, getEntities, getRegional, IRegional, getEntityByIdGeneric,
        getSectors, getBundles, getChannels, getColors, IColor, IEntityBasic} from '../../refs/global';
import * as R from 'ramda';
import {approximate, capitalize} from '@devinit/prelude';

interface IUnbundlingAidQuery {
    from_di_id?: string;
    to_di_id?: string;
    year?: number;
    sector?: string;
    bundle?: string;
    channel?: string;
}
interface IUnbundlingAidOptionsData {
    sector: string[];
    bundle: string[];
    channel: string[];
}
interface IUnBundlingAidCountries {
    to: DH.IIdNamePair[];
    from: DH.IIdNamePair[];
}

interface IUnbundlingEnitity {
    id: string;
    color?: string;
    name: string;
    type?: string;
    region?: string;
}

interface IUnbundlingAidResult extends IUnbundlingAidQuery {
    value: string;
}

export default class UnbundlingAid {
    public static getSqlQueryArgs = (args: DH.IUnbundlingAidQuery): IUnbundlingAidQuery =>
        R.omit(['groupBy', 'aidType'], args) as IUnbundlingAidQuery

    private db: IDB;
    private donorsBlackList = ['country-unspecified', 'region-unspecified', 'organisation-unspecified',
                    'arab-fund', 'afesd', 'idb-sp-fund'];
    constructor(db: any) {
        this.db = db;
    }
    public async getUnbundlingAidData(args: DH.IUnbundlingAidQuery): Promise<DH.IAidUnit[]> {
        try {
            const queryArgs: IUnbundlingAidQuery = UnbundlingAid.getSqlQueryArgs(args);
            const concept = await this.getUnbundlingAidConceptData(args.aidType);
            const table = concept.id;
            const queryStr: string = makeSqlAggregateQuery(queryArgs, args.groupBy, table);
            const raw: IUnbundlingAidResult[] = await this.db.manyCacheable(queryStr);
            const entites: IUnbundlingEnitity[] = await entitesFnMap[args.groupBy]();
            const regions: IRegional[] = await getRegional();
            const colors = await getColors();
            return raw.map((obj) => {
                const entity: IUnbundlingEnitity | undefined = entites.find(item => obj[args.groupBy] === item.id );
                if (!entity) {
                    throw new Error(`error getting unbundling aid entity
                        ${entity} \n for object ${JSON.stringify(obj)}`);
                }
                let color = entity.color || 'grey';
                if (entity.type && entity.region) {
                    const region: IRegional | undefined = getEntityByIdGeneric<IRegional>(entity.region, regions);
                    if (region && region.color) color = region ? region.color : color;
                }
                const colorObj: IColor = getEntityByIdGeneric<IColor>(color, colors);
                return {id: entity.id, value: Number(obj.value), name: entity.name, uid: shortid.generate(),
                        color: colorObj.value, year: Number(obj.year)};
           });
       } catch (error) {
           console.error(error);
           throw error;
       }
    }
    public async getUnbundlingSelectionData({aidType}): Promise<DH.IUnbundlingAidSelections> {
        try {
            const concept = await this.getUnbundlingAidConceptData(aidType);
            if (!concept) throw new Error('error getting unbundling aid concept');
            const years: number[] = R.range(Number(concept.start_year), Number(concept.end_year + 1)).reverse();
            const optionIds: IUnbundlingAidOptionsData = await this.getUnbundlingOptionsIds(aidType);
            const countries = await this.getCountries();
            const gChannels = await getChannels();
            const channels = this.getOptionsSet('channel', gChannels, optionIds);
            const gSectors = await getSectors();
            const sectors = this.getOptionsSet('sector', gSectors, optionIds);
            const gBundles = await getBundles();
            const bundles = this.getOptionsSet('bundle', gBundles, optionIds);
            return {years, ...countries, channels, sectors, bundles};
       } catch (error) {
           console.error(error);
           throw error;
       }
    }
    public async getUnbundlingAidDataTotal(args: DH.IUnbundlingAidToTalQuery): Promise<DH.IUnbundlingAidTotal> {
        const concept = await this.getUnbundlingAidConceptData(args.aidType);
        const id: string = concept.id;
        try {
            const year: number = args.year || (await getConceptAsync(`unbundling-${args.aidType}`, id)).end_year;
            const  queryArgs = {table: id, year};
            const raw: Array<{sum: string}> = await this.db.manyCacheable(sql.total, queryArgs);
            const total = approximate(raw[0].sum, 1);
            return {total, year};
        } catch (error) {
            throw error;
        }
    }
    public async getUnbundlingOptionsIds(aidType): Promise<IUnbundlingAidOptionsData> {
        const {id} = await this.getUnbundlingAidConceptData(aidType);
        const columns = ['bundle', 'channel', 'sector'];
        type IRawOptions = Array<{[column: string]: string}>;
        const optionsPromises: Array<Promise<IRawOptions>> = columns.map(async (column) => {
            const queryArgs = {table: id, column};
            const raw: IRawOptions =
                await this.db.manyCacheable(sql.options, queryArgs);
            return raw;
        }, {});
        const options: IRawOptions[]  = await Promise.all(optionsPromises);
        return options.reduce((acc, optionsObj) => {
            const key = Object.keys(optionsObj[0])[0] as string;
            const columnFields = optionsObj.map(obj => obj[key]);
            return {...acc, [key]: columnFields };
        }, {}) as IUnbundlingAidOptionsData;
    }
    private getOptionsSet(optionName, cmsOptions, dBOptionsData): IEntityBasic[] {
        const options = dBOptionsData[optionName];
        return options.map(id => {
            const option = cmsOptions.find(obj => obj.id === id);
            const name = capitalize(id.replace(/\-/g, ' '));
            if (!option) return {id, name, color: 'grey'};
            return option;
        });
    }
    private async getUnbundlingAidConceptData(aidType): Promise<IConcept> {
        const allConcepts = await getConcepts(`unbundling-${aidType}`);
        const concept = allConcepts[0];
        if (!concept) throw new Error(`${aidType} concept file missing`);
        return concept;
    }
    private async getCountries(): Promise<IUnBundlingAidCountries> {
        try {
            const entites: IEntity[] = await getEntities();
            return entites.reduce((countries: IUnBundlingAidCountries, entity) => {
            let to: any = [];
            let from: any = [];
            if (entity.donor_recipient_type === RECIPIENT || entity.donor_recipient_type === CROSSOVER
                || entity.donor_recipient_type === 'region') {
                to = R.append({id: entity.id, name: entity.name}, countries.to);
            }
            if (entity.donor_recipient_type === DONOR || entity.donor_recipient_type === MULTILATERAL
                || entity.donor_recipient_type === CROSSOVER) {
                from = R.contains(entity.id, this.donorsBlackList) ? countries.from :
                    R.append({id: entity.id, name: entity.name}, countries.from);
            }
            if (to.length && from.length) return {to, from};
            if (to.length)  return {...countries, to};
            if (from.length)  return {...countries, from};
            return countries;
            }, {to: [], from: []});
        } catch (error) {
            throw error;
        }
    }

}
