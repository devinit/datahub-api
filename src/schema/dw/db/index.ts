import {IMain, IDatabase, IOptions} from 'pg-promise';
import {dwConfig} from '../config';
import Maps from '../modules/Maps';
import UnbundlingAid from '../modules/UnbundlingAid';
import CountryProfile from '../modules/CountryProfile';
import diagnostics from './diagnostics';
import * as pgPromise from 'pg-promise';
import * as LRU from 'lru-cache';
import {queue} from '../../../lib/cache';

export interface IExtensions {
    maps: Maps;
    unbundlingAid: UnbundlingAid;
    countryProfile: CountryProfile;
    manyCacheable: (query: string, values: any) => Promise<any>;
}

const lruOpts: LRU.Options<any> = {
    max: 500,
    maxAge: 1000 * 60 * 60 * 60
};

export const dbCache: LRU.Cache<any> = LRU(lruOpts);

// pg-promise initialization options:
const options: IOptions<IExtensions> = {
    // Extending the database protocol with our custom modules
    // API: http://vitaly-t.github.io/pg-promise/global.html#event:extend
    extend: (obj: IExtensions & IDatabase<IExtensions>) => {
        obj.manyCacheable = (query: string, values?: any) => {
            const getQuery = values ? pgPromise.as.format(query, values) : query;
            if (dbCache.has(getQuery)) {
                // add to queue so that we always have freshest data
                // makes same query in 15 minutes so as to update cache
                queue(getQuery, 'dw', dbCache, obj.many);
                return Promise.resolve(dbCache.get(getQuery));
            }
            return obj.any(getQuery);
        };
        obj.maps = new Maps(obj);
        obj.countryProfile = new CountryProfile(obj);
        obj.unbundlingAid = new UnbundlingAid(obj);
    },
    // caching
    receive: (data, _result, event) => {
        // cache recieved
        if (!dbCache.has(event.query)) dbCache.set(event.query, data);
    }
};

const pgp: IMain = pgPromise(options);

// Create the database instance with extensions:

const db = pgp(dwConfig) as IDatabase<IExtensions> & IExtensions;

// Load and initialize optional diagnostics:

diagnostics.init(options);

process.on('exit', (code) => {
  // kill db
  pgp.end();
  console.log(`About to exit with code: ${code}, closed DB connection`);
});

process.on('SIGINT', () => {
    // kill db
    pgp.end();
    console.log('Ctrl-C... forced termination closed DB connection');
});

// If you ever need access to the library's root (pgp object), you can do it via db.$config.pgp
// See: http://vitaly-t.github.io/pg-promise/Database.html#.$config
export default db;
