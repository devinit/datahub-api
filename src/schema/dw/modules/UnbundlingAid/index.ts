import {IDatabase} from 'pg-promise';
import {IExtensions} from '../../db';
import {makeSqlAggregateQuery, entitesFnMap, DONOR, RECIPIENT, MULTILATERAL, CROSSOVER,
    formatNumbers} from '../utils';
import {getConceptAsync, IConcept} from '../../../cms/modules/concept';
import * as shortid from 'shortid';
import sql from './sql';
import {IEntity, getEntities, getRegional, IRegional, getEntityByIdGeneric,
        getSectors, getBundles, getChannels, getColors, IColor} from '../../../cms/modules/global';
import * as R from 'ramda';

interface IUnbundlingAidQuery {
    from_di_id?: string;
    to_di_id?: string;
    year?: number;
    sector?: string;
    bundle?: string;
    channel?: string;
}
interface IUnBundlingAidCountries {
    to: DH.IIdNamePair[];
    from: DH.IIdNamePair[];
}

interface IUnbundlingEnitity {
    id: string;
    color?: string;
    name: string;
    type?: string;
    region?: string;
}

interface IUnbundlingAidResult extends IUnbundlingAidQuery {
    value: string;
}

export default class UnbundlingAid {
    public static getSqlQueryArgs = (args: DH.IUnbundlingAidQuery): IUnbundlingAidQuery =>
        R.omit(['groupBy', 'aidType'], args) as IUnbundlingAidQuery;

    private db: IDatabase<IExtensions> & IExtensions;
    private donorsBlackList = ['country-unspecified', 'region-unspecified', 'organisation-unspecified',
                    'arab-fund', 'afesd', 'idb-sp-fund'];
    constructor(db: any) {
        this.db = db;
    }
    public async getUnbundlingAidData(args: DH.IUnbundlingAidQuery): Promise<DH.IAidUnit[]> {
        try {
            const queryArgs: IUnbundlingAidQuery = UnbundlingAid.getSqlQueryArgs(args);
            const table = this.getUnbundlingAidDataTable(args.aidType);
            const queryStr: string = makeSqlAggregateQuery(queryArgs, args.groupBy, table);
            const raw: IUnbundlingAidResult[] = await this.db.manyCacheable(queryStr, null);
            const entites: IUnbundlingEnitity[] = await entitesFnMap[args.groupBy]();
            const regions: IRegional[] = await getRegional();
            const colors = await getColors();
            return raw.map((obj) => {
                const entity: IUnbundlingEnitity | undefined = entites.find(item => obj[args.groupBy] === item.id);
                if (!entity) {
                    throw new Error(`error getting unbundling aid entity
                        ${entity} \n for object ${JSON.stringify(obj)}`);
                }
                let color = entity.color || 'grey';
                if (entity.type && entity.region) {
                    const region: IRegional | undefined = getEntityByIdGeneric<IRegional>(entity.region, regions);
                    if (region && region.color) color = region ? region.color : color;
                }
                const colorObj: IColor = getEntityByIdGeneric<IColor>(color, colors);
                return {id: entity.id, value: Number(obj.value), name: entity.name, uid: shortid.generate(),
                        color: colorObj.value, year: Number(obj.year)};
           });
       } catch (error) {
           console.error(error);
           throw error;
       }
    }
    public async getUnbundlingSelectionData({aidType}): Promise<DH.IUnbundlingAidSelections> {
        try {
            const table = this.getUnbundlingAidDataTable(aidType);
            const concept: IConcept = await getConceptAsync(`unbundling-${aidType}`, table);
            if (!concept) throw new Error('error getting unbundling aid concept');
            const years: number[] = R.range(Number(concept.start_year), Number(concept.end_year + 1)).reverse();
            const countries = await this.getCountries();
            const channels = await getChannels();
            const sectors = await getSectors();
            const allBundles = await getBundles();
            const bundles = aidType === 'oda' ? allBundles.filter(bundle => !bundle.id.includes('oof'))
                : allBundles.filter(bundle => bundle.id.includes('oof'));
            return {years, ...countries, channels, sectors, bundles};
       } catch (error) {
           console.error(error);
           throw error;
       }
    }
    public async getUnbundlingAidDataTotal(args: DH.IUnbundlingAidToTalQuery): Promise<DH.IUnbundlingAidTotal> {
        let year: number | undefined | null = args.year;
        const id: string = this.getUnbundlingAidDataTable(args.aidType);
        try {
            if (!year) {
                const concept: IConcept = await getConceptAsync(`unbundling-${args.aidType}`, id);
                year = concept.end_year || 2015;
            }
            const  queryArgs = {table: id, year};
            const raw: Array<{sum: string}> = await this.db.manyCacheable(sql.total, queryArgs);
            const total = formatNumbers(raw[0].sum, 1);
            return {total, year};
        } catch (error) {
            throw error;
        }
    }
    private getUnbundlingAidDataTable(aidType) {
        return aidType === 'oda' ? 'fact.oda_2015' : 'fact.oof_2015';
    }
    private async getCountries(): Promise<IUnBundlingAidCountries> {
        try {
            const entites: IEntity[] = await getEntities();
            return entites.reduce((countries: IUnBundlingAidCountries, entity) => {
            let to: any = [];
            let from: any = [];
            if (entity.donor_recipient_type === RECIPIENT || entity.donor_recipient_type === CROSSOVER
                || entity.donor_recipient_type === 'region') {
                to = R.append({id: entity.id, name: entity.name}, countries.to);
            }
            if (entity.donor_recipient_type === DONOR || entity.donor_recipient_type === MULTILATERAL
                || entity.donor_recipient_type === CROSSOVER) {
                from = R.contains(entity.id, this.donorsBlackList) ? countries.from :
                    R.append({id: entity.id, name: entity.name}, countries.from);
            }
            if (to.length && from.length) return {to, from};
            if (to.length)  return {...countries, to};
            if (from.length)  return {...countries, from};
            return countries;
            }, {to: [], from: []});
        } catch (error) {
            throw error;
        }
    }

}
