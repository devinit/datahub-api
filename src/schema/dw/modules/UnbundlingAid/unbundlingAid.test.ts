import * as prettyFormat from 'pretty-format';
import UnbundlingAid from '.';
import db from '../../db';

describe('Unbundling aid DW module tests', () => {
    const unbundlingAid = new UnbundlingAid(db);

    it('getting unbundling aid data of various types', async () => {
        const argsA = { aidType: 'oda', year: 2015, groupBy: 'sector', from_di_id: 'GB'};
        // const argsB = { aidType: 'oda', year: 2015, sector: 'banking-and-business', groupBy: 'to_di_id'};
        // const argsC = { aidType: 'oof', year: 2015, groupBy: 'bundle'};
        // const argsD = { aidType: 'oda', year: 2015, groupBy: 'sector'};
        const dataC = await unbundlingAid.getUnbundlingAidData(argsC);
        const dataA = await unbundlingAid.getUnbundlingAidData(argsA);
        // const dataC = await unbundlingAid.getUnbundlingAidData(argsC);
        // const dataD = await unbundlingAid.getUnbundlingAidData(argsD);
        expect(prettyFormat({dataA, dataC})).toMatchSnapshot();
    }, 100000);
    it('should create sql query args for getting data', () => {
        const argsA = {
            aidType: 'oda',
            year: 2015,
            groupBy: 'to_di_id'
        };
        const formatted = UnbundlingAid.getSqlQueryArgs(argsA);
        expect(prettyFormat(formatted)).toMatchSnapshot();
    });
    it('should return unbundling aid total', async () => {
        const totalAndYearODA = await unbundlingAid.getUnbundlingAidDataTotal({aidType: 'oda'});
        const totalAndYearOOF = await unbundlingAid.getUnbundlingAidDataTotal({aidType: 'oof'});
        expect(prettyFormat({totalAndYearODA, totalAndYearOOF})).toMatchSnapshot();
    }, 10000);
    it('getting unbundling aid selection options', async () => {
        const dataODA = await unbundlingAid.getUnbundlingSelectionData({aidType: 'oda'});
        const dataOOF = await unbundlingAid.getUnbundlingSelectionData({aidType: 'oof'});
        expect(prettyFormat({dataODA, dataOOF})).toMatchSnapshot();
    }, 100000);
    afterAll(() => {
       db.$config.pgp.end();
    });
});
