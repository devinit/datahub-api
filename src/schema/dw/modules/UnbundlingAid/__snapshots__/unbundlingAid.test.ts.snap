// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unbundling aid DW module tests getting unbundling aid data of various types 1`] = `
"Object {
  \\"dataA\\": Array [
    Object {
      \\"color\\": \\"#e8443a\\",
      \\"id\\": \\"agriculture-and-food-security\\",
      \\"name\\": \\"Agriculture and food security\\",
      \\"uid\\": \\"SJU6JmmBM\\",
      \\"value\\": 683461349.2757,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#d64278\\",
      \\"id\\": \\"banking-and-business\\",
      \\"name\\": \\"Banking and business\\",
      \\"uid\\": \\"BJxLpy7XHM\\",
      \\"value\\": 960414844.446,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#e8443a\\",
      \\"id\\": \\"education\\",
      \\"name\\": \\"Education\\",
      \\"uid\\": \\"HJZLT1m7BG\\",
      \\"value\\": 1001708266.0027188,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#f18e5e\\",
      \\"id\\": \\"environment\\",
      \\"name\\": \\"Environment\\",
      \\"uid\\": \\"HyMUaJm7rz\\",
      \\"value\\": 501812904.011,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#893f90\\",
      \\"id\\": \\"general-budget-support\\",
      \\"name\\": \\"General budget support\\",
      \\"uid\\": \\"BkmIT1mmSM\\",
      \\"value\\": 75944800,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#d85b31\\",
      \\"id\\": \\"governance-security-and-civil-society\\",
      \\"name\\": \\"Governance, security and civil society\\",
      \\"uid\\": \\"HJVU6J7XHf\\",
      \\"value\\": 1577529341.5161,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#007851\\",
      \\"id\\": \\"health\\",
      \\"name\\": \\"Health\\",
      \\"uid\\": \\"HkHL6y7mBf\\",
      \\"value\\": 1536669265.7849,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#bc2629\\",
      \\"id\\": \\"humanitarian\\",
      \\"name\\": \\"Humanitarian\\",
      \\"uid\\": \\"BJ8UpyX7BG\\",
      \\"value\\": 1942821602.524,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#5da3d9\\",
      \\"id\\": \\"industry-and-trade\\",
      \\"name\\": \\"Industry and trade\\",
      \\"uid\\": \\"SJDITJmXHM\\",
      \\"value\\": 216979058.048,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#f18e5e\\",
      \\"id\\": \\"infrastructure\\",
      \\"name\\": \\"Infrastructure\\",
      \\"uid\\": \\"SJO8ak77BM\\",
      \\"value\\": 411066389.450742,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#7b3b89\\",
      \\"id\\": \\"other\\",
      \\"name\\": \\"Other\\",
      \\"uid\\": \\"H1YLaJQmBz\\",
      \\"value\\": 2204973282.188533,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#893f90\\",
      \\"id\\": \\"other-social-services\\",
      \\"name\\": \\"Other social services\\",
      \\"uid\\": \\"BkcLTy7mHf\\",
      \\"value\\": 445825223.67,
      \\"year\\": 2015,
    },
    Object {
      \\"color\\": \\"#0089cc\\",
      \\"id\\": \\"water-and-sanitation\\",
      \\"name\\": \\"Water and sanitation\\",
      \\"uid\\": \\"S1j86kXXrf\\",
      \\"value\\": 282071261.8133,
      \\"year\\": 2015,
    },
  ],
}"
`;
